### ISSUE TITLE
Unexpected behaviour with MinDiscrepancyGreedy

### ISSUE THREAD

**Issue by 610v4nn1** [antoinedemathelin/dbal#5]
I am running the code of `MinDiscrepancyGreedy` adapted to run with pytorch and I am observing some strange behavior. In particular, the `potential` vector is always made of a single element and the algorithm always selects the point with index zero.
It could be that my adaptation introduced some bugs, but I am having hard time debugging because there are very few comments and information available in the code.

I believe this is an issue with the shapes of the matrices used in the lines before but I am currently not sure about it.
I looking into the computation of `deltas_d_plus_minus` (and lines below) because I would have expected this to be num_points x 1 but it's a 1x1 matrix.

It would be very useful if you could provide a few comments with the expected shapes of the different variables.

Below you can see the output of the debugger when running the algorithm on a toy dataset.
![Screenshot 2021-07-01 at 10 34 21](https://user-images.githubusercontent.com/52964960/124093285-fefdfc00-da57-11eb-87f6-6116b3c7e30a.png)

**Comment by antoinedemathelin** [comment]
Hi 610v4nn1,
Thank you for your interest in the `MinDiscrepancyGreedy` method. Please tell me if you found some improvements in using this method instead of KMedoids.
For the problem you have, I think the issue is in the loss function. I think the loss you use is averaging over examples, thus it returns a single element where it should return a matrix of size `(n_samples, n_samples)`. Thus `potential` will be a vector of size ``n_samples``  (with ``n_samples`` the length of X (here 30)). To correct this, you need to implement your own loss function like this:
```
def loss_fct_mae(x, y):
    return np.mean(np.abs(x - y), axis=-1)
```
Here in numpy but you can do the same with pytorch, the only think is to average or sum over the last axis.

Best,

**Comment by 610v4nn1** [comment]
Thanks for the quick answer. This is very useful. 
I had a similar intuition but I did not want to bias your answer.

There is only one detail which I don't get. I would expect potential to have length equal to the length of the tgt_index (in this case 20) and not the length of X (in this case 30).

**Comment by antoinedemathelin** [comment]
Hi 610v4nn1,
Yes, I meant indeed ``n_samples = len(tgt_index) = 20``. Sorry for the mistake.
Best,

**Comment by 610v4nn1** [comment]
Thanks a lot for the clarification :)
